syntax = "proto3";
import "message_type.proto";
import "sparql_tree.proto";
package proto_msg;


option java_package = "com.github.cristobalm.RDFEWK2C.proto";
option java_outer_classname = "RequestMsg";

message CacheRequest{
    MessageType request_type = 1;

    oneof request {
        ConnectionEndRequest connection_end_request = 2;
        CacheRunQueryAlgebra cache_run_query_algebra = 3;
        ReceiveRemainingResult receive_remaining_result = 4;
        CacheRequestSeparatePredicates cache_request_separate_predicates = 5;
        DoneWithPredicatesNotify done_with_predicates_notify = 6;
        StreamRequestTriplePattern stream_request_triple_pattern = 7;
        StreamContinueTriplePattern stream_continue_triple_pattern = 8;
    }
}
message StreamContinueTriplePattern{
    int32 channel_id = 1;
    int32 pattern_channel_id = 2;
}

message StreamRequestTriplePattern{
    int32 channel_id = 1;
    TripleNode triple_node = 2;
}

message DoneWithPredicatesNotify{
    repeated RDFTerm predicates = 1;
}

message CacheRequestSeparatePredicates{
    repeated RDFTerm predicates = 1;
}


message ReceiveRemainingResult{
    int32 id = 1;
}

message ConnectionEndRequest{
    bool end = 1;
}

message CacheRunQueryAlgebra{
    SparqlTree sparql_tree = 1;
}

message SubjectObjectPair{
    uint64 subject = 1;
    uint64 object = 2;
}

message RDFTriple {
    uint64 subject = 1;
    uint64 predicate = 2;
    uint64 object = 3;
}
